### PR TITLE
OSASINFRA-3550: openstack: ingress API & octavia ingress provider

### PR DESCRIPTION
**Pull request by EmilienM**
**What this PR does / why we need it**:

Until now, we didn't have a working Ingress on an initial deployment, and we thought we would just rely on day 2 operations to setup Ingress with e.g. MetalLB.
It turns out that we could use OpenStack Octavia (Load Balancer as a service) as an optional provider for Ingress, at least for now in dev-preview.

With this PR, we don't expect the OpenStack cloud to have Octavia enabled which is why we disable the provider by default, which means Ingress won't be working on day 1 and it's up to the user to deploy something (e.g. MetalLB).

If Octavia is used, the floating IP and the DNS record for Ingress would have to be pre-created before deploying the hosted cluster.
The new CLI would be:
```bash
hypershift create cluster openstack --openstack-ingress-provider Octavia --openstack-ingress-floating-ip <-ingress-floating-IP> (...)
```

We have some validation that check that if a FIP is provided, it's only accepted with the Octavia provider for now.
The HCP.Spec has been updated as well to allow the Ingress FIP to be stored and the Ingress provider (default to `None` now).

In the future, we might change the default and even remove the Octavia provider if we find a better way (e.g. re-using on-prem LB, based on Keepalived).
But by keeping it "disabled" by default ensure some sort of backward compatibility even if OpenStack is dev-preview now.

### CODE REVIEW

File: support/openstackutil/ingress.go
Comment (MaysaMacedo) [inline support/openstackutil/ingress.go:22]: Based on the comment above, shouldn't it be like this:
```suggestion
	if ingressFloatingIP != "" && ingressProvider != hyperv1.OpenStackIngressProviderOctavia {
```

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go
Comment (MaysaMacedo) [inline control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:937]: why do we need to validate it again if it was already validate at `cmd/cluster/openstack/create.go`?
Is it because the hostedcluster could be created without cmdline?

Comment (EmilienM) [reply control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:937]: yeah exactly, I thought it'll be nice to avoid controller errors in case someone creates the cluster via manifests.

File: support/openstackutil/ingress.go
Comment (EmilienM) [reply support/openstackutil/ingress.go:22]: at this point we know `ingressFloatingIP` is not empty, see check 1 and 3 before.

**Comment by EmilienM** [comment]
/test e2e-azure
resting Azure but it seems unstable

**Comment by EmilienM** [comment]
/test e2e-openstack
to ensure there is no regression on default deployments

File: api/hypershift/v1beta1/hostedcluster_types.go
Comment (EmilienM) [inline api/hypershift/v1beta1/hostedcluster_types.go:1921]: I guess I could have add more docs on this one.
I'll do it in a next PR if people don't mind, unless I'm asked to change something else, then I'll do it in batch.
Thanks

File: cmd/cluster/openstack/testdata/zz_fixture_TestCreateCluster_wrong_ingress_provider_name.yaml
Comment (EmilienM) [inline cmd/cluster/openstack/testdata/zz_fixture_TestCreateCluster_wrong_ingress_provider_name.yaml:1]: that file is wrong I need to remove it

**Comment by EmilienM** [comment]
/retest-required
/test e2e-openstack
CI glitch on the registry

**Comment by mdbooth** [comment]
I'd prefer to see https://github.com/openshift/release/pull/55115 merged first so we can the e2e tests passing.

**Comment by EmilienM** [comment]
> I'd prefer to see [openshift/release#55115](https://github.com/openshift/release/pull/55115) merged first so we can the e2e tests passing.

fair

**Comment by EmilienM** [comment]
I confirm that we have no regression with that PR:

```
    util.go:324: ClusterVersionProgressing=True: ClusterOperatorNotAvailable(Unable to apply 4.17.0-0.ci.test-2024-08-02-142747-ci-op-xv0zc09t-latest: the cluster operator console is not available)
```
like before.

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go
Comment (csrwng) [inline control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:938]: seems that should return here

**Review by csrwng — COMMENTED** [review]
Just one comment

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go
Comment (EmilienM) [reply control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:938]: done

**Comment by EmilienM** [comment]
/hold for CI
I'll hold cancel once we have positive results.

**Comment by EmilienM** [comment]
/test e2e-openstack
With the recent changes in openshift/release, we should make progress to have Ingress working in the CI job.

**Comment by EmilienM** [comment]
OpenStack CI will fail until we get https://github.com/openshift/release/pull/55160 (at least). let's wait and retest once ready.

**Comment by EmilienM** [comment]
Status of this PR:
* I want to get the aws job green to avoid we don't break anything
* once https://github.com/openshift/release/pull/55196 has merged, we can run e2e-openstack and hopefully get good results. If not we'll iterate.

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/ingress/reconcile.go
Comment (sjenning) [inline control-plane-operator/hostedclusterconfigoperator/controllers/resources/ingress/reconcile.go:174]: This is erasing the exist spec leading to a hot-reconcile loop
```
{"level":"info","ts":"2024-08-05T22:41:07Z","msg":"WARNING: Object got updated more than one time without a no-op update, this indicates hypershift incorrectly reverting defaulted values","type":"*v1.Service","name":"openshift-ingress/octavia-ingress","diff":"  &v1.Service{
  	TypeMeta:   {Kind: "Service", APIVersion: "v1"},
  	ObjectMeta: {Name: "octavia-ingress", Namespace: "openshift-ingress", UID: "47e7e65e-63d9-4291-b2f3-32fec7b5f5dd", ResourceVersion: "1306", ...},
  	Spec: v1.ServiceSpec{
  		Ports: []v1.ServicePort{
  			{
  				... // 3 identical fields
  				Port:       80,
  				TargetPort: {IntVal: 80},
- 				NodePort:   31330,
+ 				NodePort:   0,
  			},
  			{
  				... // 3 identical fields
  				Port:       443,
  				TargetPort: {IntVal: 443},
- 				NodePort:   30416,
+ 				NodePort:   0,
  			},
  		},
  		Selector:  {"ingresscontroller.operator.openshift.io/deployment-ingresscontroller": "default"},
  		ClusterIP: "172.31.230.21",
  		... // 5 identical fields
  		LoadBalancerSourceRanges:      nil,
  		ExternalName:                  "",
- 		ExternalTrafficPolicy:         "Cluster",
+ 		ExternalTrafficPolicy:         "",
  		HealthCheckNodePort:           0,
  		PublishNotReadyAddresses:      false,
  		SessionAffinityConfig:         nil,
  		IPFamilies:                    {"IPv4"},
  		IPFamilyPolicy:                &"SingleStack",
- 		AllocateLoadBalancerNodePorts: &true,
+ 		AllocateLoadBalancerNodePorts: nil,
  		LoadBalancerClass:             nil,
- 		InternalTrafficPolicy:         &"Cluster",
+ 		InternalTrafficPolicy:         nil,
  		TrafficDistribution:           nil,
  	},
  	Status: {},
  }
","semanticDeepEqual":false,"updateCount":10}
```
The spec should be reconciled field by field, not rebuild from scratch.

**Review by sjenning — CHANGES_REQUESTED** [review]
/hold

**Comment by EmilienM** [comment]
Note: for testing purpose, I cherry picked the commits from https://github.com/openshift/hypershift/pull/4492 until they are merged. This PR is still on hold for now, until we get green CI on e2e-openstack.

**Comment by EmilienM** [comment]
@sjenning I removed the 2 commits from the other PR since I got better results, this PR can now land I think. Please add the LGTM again.
Thanks

/hold cancel

**Pull request merged**
